applications/DEMApplication/python_scripts/DEM_analysis_stage.py

```diff
@@ -0,0 +1,647 @@
+            del self.analytic_model_part
+
+        del self.KRATOSprint
+        del self.all_model_parts
+        del self.demio
+        del self.procedures
+        del self.creator_destructor
+        del self.dem_fem_search
+        del self.solver
+        del self.DEMFEMProcedures
+        del self.post_utils
+        del self.cluster_model_part
+        del self.rigid_face_model_part
+        del self.spheres_model_part
+        del self.DEM_inlet_model_part
+        del self.mapping_model_part
+
+        if self.DEM_parameters["dem_inlet_option"].GetBool():
+            del self.DEM_inlet
+
+    def SetGraphicalOutput(self):
+        self.demio = DEM_procedures.DEMIo(self.model, self.DEM_parameters, self.post_path, self.all_model_parts)
+        if self.DEM_parameters["post_vtk_option"].GetBool():
+            import dem_vtk_output
+            self.vtk_output = dem_vtk_output.VtkOutput(self.main_path, self.problem_name, self.spheres_model_part, self.rigid_face_model_part)
+
+    def GraphicalOutputInitialize(self):
+        self.demio.Initialize(self.DEM_parameters)
+
+        #-------------os.chdir(self.post_path)
+        self.demio.InitializeMesh(self.all_model_parts)
+
+    def PrintResultsForGid(self, time):
+        if self.solver.poisson_ratio_option:
+            self.DEMFEMProcedures.PrintPoisson(self.spheres_model_part, self.DEM_parameters, "Poisson_ratio.txt", time)
+
+        if self.DEM_parameters["PostEulerAngles"].GetBool():
+            self.post_utils.PrintEulerAngles(self.spheres_model_part, self.cluster_model_part)
+
+        self.demio.ShowPrintingResultsOnScreen(self.all_model_parts)
+
+        #------os.chdir(self.data_and_results)
+        self.demio.PrintMultifileLists(time, self.post_path)
+        self.solver.PrepareElementsForPrinting()
+        if self.DEM_parameters["ContactMeshOption"].GetBool():
+            self.solver.PrepareContactElementsForPrinting()
+
+        #os.chdir(self.post_path)
+
+        self.demio.PrintResults(self.all_model_parts, self.creator_destructor, self.dem_fem_search, time, self.bounding_box_time_limits)
+        #------os.chdir(self.main_path)
+        if "post_vtk_option" in self.DEM_parameters.keys():
+            if self.DEM_parameters["post_vtk_option"].GetBool():
+                self.vtk_output.WriteResults(self.time)
+
+    def GraphicalOutputFinalize(self):
+        self.demio.FinalizeMesh()
+        self.demio.CloseMultifiles()
+
+
+    #these functions are needed for coupling, so that single time loops can be done
+
+    def InitializeTime(self):
+        self.step = 0
+        self.time = 0.0
+        self.time_old_print = 0.0
+
+    def UpdateTimeParameters(self):
+        self.InitializeSolutionStep()
+        self.step, self.time = self._GetSolver().AdvanceInTime(self.step, self.time)
+        self.DEMFEMProcedures.UpdateTimeInModelParts(self.all_model_parts, self.time, self.solver.dt, self.step)
+
+    def FinalizeSingleTimeStep(self):
```

Comment:
I think this should be FinalizeSolutionStep

Reply:
FinalizeSolutionStep is there. This function is only used by the DEM-FEM coupling.

Reply:
hm, then this also breaks the interface for generic coupling
cannot you merge those two functions?

Reply:
Yes, but I have to talk to the people in charge of the DEM-FEM. I think it is best to leave it at this point for this PR to avoid breaking their code. I take note of all your comments, though.